spec/controllers/churn_controller_spec.rb

```diff
@@ -0,0 +1,203 @@
+        allow(service_instance).to receive(:fetch_churn_data).and_return(nil)
+
+        get :show
+      end
+
+      it "uses InertiaRails.optional for lazy loading churn_data" do
+        expect(InertiaRails).to receive(:optional).and_call_original
+        get :show
+      end
+    end
+
+    context "when user has no subscription products" do
+      let!(:product) { create(:product, user: user) }
+
+      it "renders successfully and indicates no subscription products" do
+        get :show
+        expect(response).to be_successful
+        expect(inertia.props[:churn_props]).to include(
+          has_subscription_products: false
+        )
+      end
+
+      it "returns empty products array" do
+        get :show
+        expect(inertia.props[:churn_props][:products]).to be_empty
+      end
+    end
+
+    context "with date parameters" do
+      let(:from_date) { "2024-01-01" }
+      let(:to_date) { "2024-01-31" }
+
+      it "passes date parameters to service" do
+        expect(CreatorAnalytics::Churn).to receive(:new).with(
+          user: user,
+          params: hash_including(from: from_date, to: to_date)
+        ).and_call_original
+
+        get :show, params: { from: from_date, to: to_date }
+      end
+    end
+
+    context "with product parameters" do
+      let(:product_ids) { [product.id] }
+
+      it "passes product parameters to service" do
+        expect(CreatorAnalytics::Churn).to receive(:new).with(
+          user: user,
+          params: instance_of(ActionController::Parameters)
+        ).and_call_original
+
+        get :show, params: { products: product_ids }
+      end
+    end
+
+    context "with Rails cache for large sellers" do
+      let!(:large_seller) { create(:large_seller, user: user) }
+
+      it "uses cache when churn_data is requested with partial reload" do
+        allow(Rails.cache).to receive(:fetch).and_call_original
+        allow(Rails.cache).to receive(:fetch).with(
+          match(/seller_daily_churn_metrics:#{user.id}/),
+          expires_in: 24.hours
+        ).and_return({})
+
+        get :show, params: { only: ["churn_data"] }
+        expect(response).to be_successful
+      end
+
+      it "returns cached data when available" do
+        cached_data = { customer_churn_rate: 5.5, churned_subscribers: 10 }
+        allow(Rails.cache).to receive(:fetch).and_call_original
+        allow(Rails.cache).to receive(:fetch).with(
+          match(/seller_daily_churn_metrics:#{user.id}/),
+          expires_in: 24.hours
+        ).and_return(cached_data)
+
+        get :show, params: { only: ["churn_data"] }
+        expect(response).to be_successful
+      end
+    end
+
+    context "without caching" do
+      before do
+        allow(LargeSeller).to receive(:where).and_return(double(exists?: false))
+      end
+
+      it "calculates data in real-time when churn_data is requested" do
+        allow(Rails.cache).to receive(:fetch).and_call_original
```

Comment:
If I'm not mistaken, this mocks the call and expects it. Can you check?
```suggestion
        expect(Rails.cache).to receive(:fetch).and_call_original
```